# Team 5: Sales of Summer Clothes in E-commerce Wish 
Columbia University - SQL & Relational Databases 

By Han Xiao, Jiahui Ren, Katherine Li, Shan Lu, Yuxiang Liu

First, import necessary packages

In [1]:
import numpy as np 
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sqlalchemy import create_engine

# Dataset inspection

Load the csv file in a dataframe, df:

In [3]:
wish = pd.read_csv('/Users/lushan/Desktop/summer-products-with-rating-and-performance_2020-08.csv')
df = wish

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1573 entries, 0 to 1572
Data columns (total 43 columns):
title                           1573 non-null object
title_orig                      1573 non-null object
price                           1573 non-null float64
retail_price                    1573 non-null int64
currency_buyer                  1573 non-null object
units_sold                      1573 non-null int64
uses_ad_boosts                  1573 non-null int64
rating                          1573 non-null float64
rating_count                    1573 non-null int64
rating_five_count               1528 non-null float64
rating_four_count               1528 non-null float64
rating_three_count              1528 non-null float64
rating_two_count                1528 non-null float64
rating_one_count                1528 non-null float64
badges_count                    1573 non-null int64
badge_local_product             1573 non-null int64
badge_product_quality           1573 non-nul

As we can see above, this dataset has 43 columns and 1573 rows. Within all of the columns, 19 of them have been assigned "object" datatype, 15 columns have been assigned "int64" datatype, and 9 columns have been assigned "float64" datatype. Also, not all the columns have 1573 rows so the following columns contain NULL values: "rating_five_count", "rating_four_count", "rating_three_count", "rating_two_count", "rating_one_count", "product_color", "product_variation_size_id", "has_urgency_banner", "urgency_text", "origin_country", "merchant_name", "merchant_info_subtitle", and "merchant_profile_picture". We will need to take this in to account when designing the database tables. 



# Create database tables

 Before running the commands below, we have created a project database in pgAdmin.

In [3]:
# Pass the connection string to a variable, conn_url
conn_url = 'postgresql://postgres:86368643@localhost/project'

# Create an engine that connects to PostgreSQL server
engine = create_engine(conn_url)

# Establish a connection
connection = engine.connect()

# Pass the SQL statements that create all tables
stmt = """
CREATE TABLE merchant (
	merchant_id varchar(1000),
	merchant_title varchar(1000) NOT NULL,
	merchant_name   varchar(1000) NOT NULL,
	merchant_info_subtitle  varchar(500) NOT NULL,
	merchant_rating_count int NOT NULL,
	merchant_rating numeric NOT NULL,
	merchant_has_profile_picture int NOT NULL,
	merchant_profile_picture varchar(500) NOT NULL,
	PRIMARY KEY(merchant_id),
	CHECK (merchant_has_profile_picture IN (0,1))
	);
	

CREATE TABLE product (
	product_id    varchar(100),
	price numeric(12,2) NOT NULL,
	title_ori varchar(1000) NOT NULL,
	product_picture varchar(1000),
	product_url varchar(1000) NOT NULL,
	units_sold int NOT NULL,
	origin_country varchar(50) NOT NULL,
	retail_price numeric(12,2) NOT NULL,
	inventory_total int NOT NULL,
	uses_ad_boost int NOT NULL,
	theme varchar(100),
	badges_count int not null,
	merchant_id varchar(1000) NOT NULL,
	PRIMARY KEY(product_id),
	foreign key (merchant_id) references merchant(merchant_id),
	CHECK (uses_ad_boost IN (0,1))
	);
	
CREATE TABLE currency (
	currency_id int,
	currency_name varchar(50) NOT NULL,
	PRIMARY KEY(currency_id)
	);
	
CREATE TABLE product_currency (
	product_id varchar(100),
	currency_id int,
	currency_buyer varchar(50),
	PRIMARY KEY(product_id,currency_id),
	foreign key (product_id) references product(product_id) on delete cascade,
	foreign key (currency_id) references currency(currency_id)
	);

CREATE TABLE shipping_option (
	shipping_option_id    	int,
	shipping_option_name varchar(100) NOT NULL,
	PRIMARY KEY(shipping_option_id)
	);
	
CREATE TABLE product_shipping (
	product_id varchar(100),
	shipping_option_id    	int,
	shipping_option_price int NOT NULL,
	shipping_is_express int NOT NULL,
	countries_shipped_to int NOT NULL,
	PRIMARY KEY(product_id, shipping_option_id),
	foreign key (product_id) references product(product_id) on delete cascade,
	foreign key (shipping_option_id) references shipping_option(shipping_option_id),
	CHECK (shipping_is_express IN (0,1))
	);
	
CREATE TABLE color (
	color_id int,
	color_name    	varchar(50) NOT NULL,
	PRIMARY KEY(color_id)
	);
	
CREATE TABLE product_color (
	product_id varchar(100),
	color_id int,
	product_color    varchar(50) NOT NULL,
	PRIMARY KEY(product_id,color_id),
	foreign key (product_id) references product(product_id) on delete cascade,
	foreign key (color_id) references color(color_id)
	);	
	
CREATE TABLE tag (
	tag_id int,
	tag_name    	varchar(50) NOT NULL,
	PRIMARY KEY(tag_id)
	);
	
CREATE TABLE product_tag (
	product_id varchar(100),
	tag_id int,
	tags    	varchar(50) NOT NULL,
	PRIMARY KEY(product_id,tag_id),
	foreign key (product_id) references product(product_id) on delete cascade,
	foreign key (tag_id) references tag(tag_id)
	);
CREATE TABLE badge(
	badge_id int,
	badge_name varchar(50) NOT NULL,
	PRIMARY KEY(badge_id)
);	
create table product_badge(
	product_id varchar(100),
	badge_id int,
	foreign key (product_id) references product(product_id) on delete cascade,
	foreign key (badge_id) references badge(badge_id)
);


CREATE TABLE size_inventory (
	product_id varchar(100),
	product_variation_size_id varchar(10),
	product_variation_inventory int NOT NULL,
	PRIMARY KEY(product_id,product_variation_size_id),
	foreign key (product_id) references product(product_id) on delete cascade
);

CREATE TABLE product_rating (
	product_id varchar(100),
    rating int NOT NULL,
	rating_one_count int NOT NULL,
	rating_two_count int NOT NULL,
	rating_three_count int NOT NULL,
	rating_four_count int NOT NULL,
	rating_five_count int NOT NULL,
	PRIMARY KEY(product_id),
	foreign key (product_id) references product(product_id) on delete cascade
);

CREATE TABLE product_urgency (
	product_id varchar(100),
	has_urgency_banner int NOT NULL,
	urgency_text varchar(100),
	PRIMARY KEY(product_id),
	foreign key (product_id) references product(product_id) on delete cascade,
	CHECK (has_urgency_banner IN (0,1))
);

"""

In [4]:
# Execute the statement to create tables
connection.execute(stmt)

# Extract, Transform and Load (ETL)

With the database and all tables created (3NF), it is now time to extract, transform and load (ETL) the dataset into the database. In order to do so we will have to perform several data transformations on the loaded dataframe, df in order to create all primary keys and maintain proper relationships. We implement the ETL process according to the 15 tables that we have created above. 

1. Merchant Table  
First, we check NULL values in merchant table and replace all NULL values with 'Unknown'. Then, all merchant_id in the dataset are unique, we remove duplicates in the merchant_id column. We can now easily create a subset of df corresponding to the merchant database table. And it is now possible to load merchant data to the database.


In [6]:
# check null values 
df.isnull().any()

# drop duplicates
df = df.drop_duplicates(['merchant_id'])

# replace null values with 'Unknown'
df['merchant_name'] = df['merchant_name'].replace(np.nan,'Unknown')
df['merchant_info_subtitle'] = df['merchant_info_subtitle'].replace(np.nan,'Unknown')
df['merchant_profile_picture'] = df['merchant_profile_picture'].replace(np.nan,'Unknown')

# We can now easily create a subset of df corresponding to the merchant database table.
merchant_df = df[['merchant_id','merchant_title','merchant_name','merchant_info_subtitle','merchant_rating_count','merchant_rating','merchant_has_profile_picture','merchant_profile_picture']]

# Let's take a look:
merchant_df.head(10)

merchant_id          merchant_title  \
0  595097d6a26f6e070cb878d1                zgrdejia   
1  56458aa03a698c35c9050988               SaraHouse   
2  5d464a1ffdf7bc44ee933c65                  hxt520   
3  58cfdefdacb37b556efdff7c                allenfan   
4  5ab3b592c3911a095ad5dadb         youngpeopleshop   
5  5e4b9c3801ba9d210036fc5a           zhoulinglinga   
6  5652f4053a698c76dc9a3f37  Unique Li Fashion Shop   
7  5d45349676befe65691dcfbb                 So Band   
8  5d42980e8388970d32294ddc         chenxiangjunjun   
9  5ba2251b4315d12ebce873fa           Luowei clothe   

                                       merchant_name  \
0                                           zgrdejia   
1                                          sarahouse   
2                                             hxt520   
3                                           allenfan   
4                                        happyhorses   
5                                      zhoulinglinga   
6  uniquelifashionshopbb657bfe91d211e598c7063a14d...   
7                                             soband   
8                                    chenxiangjunjun   
9                                       luoweiclothe   

              merchant_info_subtitle  merchant_rating_count  merchant_rating  \
0                        (568 notes)                    568         4.128521   
1  83 % avis positifs (17,752 notes)                  17752         3.899673   
2     86 % avis positifs (295 notes)                    295         3.989831   
3                     (23,832 notes)                  23832         4.020435   
4  85 % avis positifs (14,482 notes)                  14482         4.001588   
5      75 % avis positifs (65 notes)                     65         3.507692   
6  86 % avis positifs (10,194 notes)                  10194         4.076516   
7                        (342 notes)                    342         3.681287   
8     82 % avis positifs (330 notes)                    330         3.803030   
9   85 % avis positifs (5,534 notes)                   5534         3.999819   

   merchant_has_profile_picture  \
0                             0   
1                             0   
2                             0   
3                             0   
4                             0   
5                             0   
6                             1   
7                             0   
8                             0   
9                             0   

                            merchant_profile_picture  
0                                            Unknown  
1                                            Unknown  
2                                            Unknown  
3                                            Unknown  
4                                            Unknown  
5                                            Unknown  
6  https://s3-us-west-1.amazonaws.com/sweeper-pro...  
7                                            Unknown  
8                                            Unknown  
9                                            Unknown

In [7]:
# load merchant table into database 
merchant_df.to_sql(name='merchant', con=engine, if_exists='append', index=False)

2. Product Table  
First, we check NULL values in the product table and replace all NULL values with 'Unknown'. Since product_id in is unique, we drop duplicates in product_id. We can now easily create a subset of df corresponding to the product database table. And it is now possible to load product data to the database.

In [8]:
# check null values 
df.isnull().any() 

# replace null values 
df['origin_country'] = df['origin_country'].replace(np.nan,'Unknown')

# drop duplicates
df = df.drop_duplicates(['product_id'])

# We can now create a subset of df corresponding to the product database table.
product_df = df[['product_id','price', 'retail_price', 'origin_country', 'product_url','product_picture','inventory_total','uses_ad_boosts','theme','merchant_id','badges_count','units_sold','title_orig']]

# rename columns 
product_df = product_df.rename(columns={"title_orig": "title_ori","uses_ad_boosts":"uses_ad_boost"})

# let's take a look 
product_df.head(10)

product_id  price  retail_price origin_country  \
0  5e9ae51d43d6a96e303acdb0  16.00            14             CN   
1  58940d436a0d3d5da4e95a38   8.00            22             CN   
2  5ea10e2c617580260d55310a   8.00            43             CN   
3  5cedf17ad1d44c52c59e4aca   8.00             8             CN   
4  5ebf5819ebac372b070b0e70   2.72             3             CN   
5  5ec645bafd107a02279c8c54   3.92             9             CN   
6  5c63a337d5e2ce4bbb3152cf   7.00             6             CN   
7  5e0ae5ebc2efb76ccf0a3391  12.00            11             CN   
8  5e6f1fb7fe4a5bb4b8bf36e5  11.00            84             CN   
9  5ccfaf238a8d535cec2dfb47   5.78            22             CN   

                                       product_url  \
0  https://www.wish.com/c/5e9ae51d43d6a96e303acdb0   
1  https://www.wish.com/c/58940d436a0d3d5da4e95a38   
2  https://www.wish.com/c/5ea10e2c617580260d55310a   
3  https://www.wish.com/c/5cedf17ad1d44c52c59e4aca   
4  https://www.wish.com/c/5ebf5819ebac372b070b0e70   
5  https://www.wish.com/c/5ec645bafd107a02279c8c54   
6  https://www.wish.com/c/5c63a337d5e2ce4bbb3152cf   
7  https://www.wish.com/c/5e0ae5ebc2efb76ccf0a3391   
8  https://www.wish.com/c/5e6f1fb7fe4a5bb4b8bf36e5   
9  https://www.wish.com/c/5ccfaf238a8d535cec2dfb47   

                                     product_picture  inventory_total  \
0  https://contestimg.wish.com/api/webimage/5e9ae...               50   
1  https://contestimg.wish.com/api/webimage/58940...               50   
2  https://contestimg.wish.com/api/webimage/5ea10...               50   
3  https://contestimg.wish.com/api/webimage/5cedf...               50   
4  https://contestimg.wish.com/api/webimage/5ebf5...               50   
5  https://contestimg.wish.com/api/webimage/5ec64...               50   
6  https://contestimg.wish.com/api/webimage/5c63a...               50   
7  https://contestimg.wish.com/api/webimage/5e0ae...               50   
8  https://contestimg.wish.com/api/webimage/5e6f1...               50   
9  https://contestimg.wish.com/api/webimage/5ccfa...               50   

   uses_ad_boost   theme               merchant_id  badges_count  units_sold  \
0              0  summer  595097d6a26f6e070cb878d1             0         100   
1              1  summer  56458aa03a698c35c9050988             0       20000   
2              0  summer  5d464a1ffdf7bc44ee933c65             0         100   
3              1  summer  58cfdefdacb37b556efdff7c             0        5000   
4              1  summer  5ab3b592c3911a095ad5dadb             0         100   
5              0  summer  5e4b9c3801ba9d210036fc5a             0          10   
6              0  summer  5652f4053a698c76dc9a3f37             0       50000   
7              0  summer  5d45349676befe65691dcfbb             0        1000   
8              1  summer  5d42980e8388970d32294ddc             0         100   
9              0  summer  5ba2251b4315d12ebce873fa             0        5000   

                                           title_ori  
0  2020 Summer Vintage Flamingo Print  Pajamas Se...  
1   Women's Casual Summer Sleeveless Sexy Mini Dress  
2  2020 New Arrival Women Spring and Summer Beach...  
3  Hot Summer Cool T Shirt for Women Fashion Tops...  
4  Women Summer Shorts Lace Up Elastic Waistband ...  
5  Plus Size Summer Women Casual Sleeveless Rompe...  
6  Women Fashion Loose Lace Blouse V Neck Bat Sle...  
7  Women's Baggy Tunic Dress Summer Dress Denim D...  
8  Women's Summer Casual Dress Fashion Short Slee...  
9  Summer Women Plus Size Casual Loose V Neck Sho...

In [9]:
# load product table into database  
product_df.to_sql(name='product', con=engine, if_exists='append', index=False)


3. Currency Table  
First, we create a dataframe called currency_df by selecting column "currency_buyer" and drop duplicates in this columns. Since all currency in the dataset are unique, we can create a new column with incrementing integer numbers for currency_id. To better understand the table, we rename column "currency_buyer" as "currency_name". Then, we load currency table to the database. 

In [10]:
# create a dataframe 
currency_df = pd.DataFrame(df['currency_buyer'])

# drop duplicates
currency_df = currency_df.drop_duplicates()

# assign unique id 
currency_df.insert(0, 'currency_id', range(1, 1 + len(currency_df)))

# rename column
currency_df = currency_df.rename(columns = {'currency_buyer':'currency_name'}) 

# let's take a look
currency_df.head(10)


currency_id currency_name
0            1           EUR

In [11]:
# check currency table 
currency_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1 entries, 0 to 0
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   currency_id    1 non-null      int64 
 1   currency_name  1 non-null      object
dtypes: int64(1), object(1)
memory usage: 24.0+ bytes


In [12]:
# load currency table into database  

currency_df.to_sql(name='currency', con=engine, if_exists='append', index=False)



4. Product_urgency Table     
First, we check NULL values in columns "has_urgency_banner" and "urgency_text" and replace all NULL values with 0. Then, we create a dataframe called "product_urgency_df" by selecting columns "product_id","has_urgency_banner", and "urgency_text". Next, we we load product_urgency table to the database.

In [13]:
# check null values
df[df['has_urgency_banner'].isnull()].head()
df[df['urgency_text'].isnull()].head()

# replace null values
df['has_urgency_banner'] = df['has_urgency_banner'].replace(np.nan, 0)
df['urgency_text'] = df['urgency_text'].replace(np.nan, '')

# create a dataframe 
product_urgency_df = pd.DataFrame(df[['product_id','has_urgency_banner', 'urgency_text']])

# let's take a look
product_urgency_df.head(10)


product_id  has_urgency_banner        urgency_text
0  5e9ae51d43d6a96e303acdb0                 1.0  Quantité limitée !
1  58940d436a0d3d5da4e95a38                 1.0  Quantité limitée !
2  5ea10e2c617580260d55310a                 1.0  Quantité limitée !
3  5cedf17ad1d44c52c59e4aca                 0.0                    
4  5ebf5819ebac372b070b0e70                 1.0  Quantité limitée !
5  5ec645bafd107a02279c8c54                 0.0                    
6  5c63a337d5e2ce4bbb3152cf                 0.0                    
7  5e0ae5ebc2efb76ccf0a3391                 0.0                    
8  5e6f1fb7fe4a5bb4b8bf36e5                 1.0  Quantité limitée !
9  5ccfaf238a8d535cec2dfb47                 0.0

In [14]:
# load product_urgency into database  
product_urgency_df.to_sql(name='product_urgency', con=engine, if_exists='append', index=False)

5. Product_rating   
First, we check NULL values and replace null values with 0. Then, we create a dataframe called "product_rating_df" by selecting columns "product_id","rating","rating_one_count","rating_two_count","rating_three_count", and "rating_four_count". Next, we load product_rating in database. 

In [15]:
# replace null values 
df['rating_five_count'] = df['rating_five_count'].replace(np.nan, 0)
df['rating_four_count'] = df['rating_four_count'].replace(np.nan, 0)
df['rating_three_count'] = df['rating_three_count'].replace(np.nan, 0)
df['rating_two_count'] = df['rating_two_count'].replace(np.nan, 0)
df['rating_one_count'] = df['rating_one_count'].replace(np.nan, 0)

# create a dataframe 
product_rating_df = pd.DataFrame(df[['product_id','rating','rating_one_count','rating_two_count','rating_three_count','rating_four_count','rating_five_count']])

# let's take a look 
product_rating_df.head(10)                                

product_id  rating  rating_one_count  rating_two_count  \
0  5e9ae51d43d6a96e303acdb0    3.76               9.0               1.0   
1  58940d436a0d3d5da4e95a38    3.45            1077.0             644.0   
2  5ea10e2c617580260d55310a    3.57               3.0               0.0   
3  5cedf17ad1d44c52c59e4aca    4.03              36.0              42.0   
4  5ebf5819ebac372b070b0e70    3.10               6.0               2.0   
5  5ec645bafd107a02279c8c54    5.00               0.0               0.0   
6  5c63a337d5e2ce4bbb3152cf    3.84             757.0             490.0   
7  5e0ae5ebc2efb76ccf0a3391    3.76              31.0              18.0   
8  5e6f1fb7fe4a5bb4b8bf36e5    3.47               3.0               1.0   
9  5ccfaf238a8d535cec2dfb47    3.60             112.0              68.0   

   rating_three_count  rating_four_count  rating_five_count  
0                10.0                8.0               26.0  
1              1118.0             1027.0             2269.0  
2                 2.0                4.0                5.0  
3                87.0              119.0              295.0  
4                 2.0                4.0                6.0  
5                 0.0                0.0                1.0  
6               971.0             1352.0             3172.0  
7                61.0               56.0              120.0  
8                 3.0                2.0                6.0  
9                92.0              128.0              287.0

In [16]:
# load product_rating into sql
product_rating_df.to_sql(name='product_rating', con=engine, if_exists='append', index=False)

6. Shipping_option Table  
First, we create a dataframe called "shipping_option_df" by selecting columns "shipping_option_name". Then we drop duplicates. Since all shipping_option in the dataset are unique, we can create a new column with incrementing integer numbers for shipping_option_id. Next, we load shipping_option in database.

In [17]:
# create a dataframe 
shipping_option_df = pd.DataFrame(wish['shipping_option_name'])

# drop duplicates
shipping_option_df = shipping_option_df.drop_duplicates()

# assign unique id 
shipping_option_df.insert(0, 'shipping_option_id', range(1, 1 + len(shipping_option_df)))

# let's take a look
shipping_option_df.head(100)

shipping_option_id     shipping_option_name
0                      1       Livraison standard
12                     2             Envio Padrão
34                     3         Standart Gönderi
42                     4          Standardversand
54                     5             Envío normal
79                     6      Standardowa wysyłka
124                    7            الشحن القياسي
133                    8       Expediere Standard
152                    9      การส่งสินค้ามาตรฐาน
177                   10        Standard Shipping
492                   11  ការដឹកជញ្ជូនតាមស្តង់ដារ
661                   12        Livraison Express
735                   13     Стандартная доставка
867                   14      Spedizione standard
1061                  15       Ekspresowa wysyłka

In [18]:
# load shipping_optionin into database 
shipping_option_df.to_sql(name='shipping_option', con=engine, if_exists='append', index=False)

7. Size_inventory Table   
First, we unify all the size into the same format, such as replace "S." into "S". Then we write a function to count amount based on different size. We replace NULL values with "other". Next, we create a dataframe called " size_inventory_df" by selecting "product_id","product_variation_size_id", and "product_variation_inventory". Last, we load size_inventory table in database. 

In [19]:
# unify size into same format
var = df['product_variation_size_id'].value_counts()
var[var > 1]
df['product_variation_size_id'] = df['product_variation_size_id'].replace('S.', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('XS.', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('M.', 'M')

df['product_variation_size_id'] = df['product_variation_size_id'].replace('Size S', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('Size-XS', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('SIZE XS', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('Size-S', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('Size4XL', 'XL')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('size S', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('Size M', 'M')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('Size -XXS', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('SIZE-XXS', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('Size S.', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('s', 'S')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('SizeL', 'L')

df['product_variation_size_id'] = df['product_variation_size_id'].replace('5XL', 'XL')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('4XL', 'XL')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('3XL', 'XL')
df['product_variation_size_id'] = df['product_variation_size_id'].replace('2XL', 'XL')

# function to count amount
var = df['product_variation_size_id'].value_counts()
var[var > 1]

def var(name):
    if name == 'XXXS' \
    or name == 'XXS' \
    or name == 'XS' \
    or name == 'S' \
    or name == 'M' \
    or name == 'L' \
    or name == 'XL' \
    or name == 'XXL' \
    or name == 'XXXXL' \
    or name == 'XXXXXL':
        return name
    else:
        return "OTHER"
    
# rename columns     
df['product_variation_size_id'] = df['product_variation_size_id'].replace(np.nan, 'OTHER')
df['product_variation_size_id'] = df['product_variation_size_id'].apply(var)

# create a dataframe 
size_inventory_df = pd.DataFrame(df[['product_id','product_variation_size_id', 'product_variation_inventory']])

# let's take a look 
size_inventory_df.head(10)

product_id product_variation_size_id  \
0  5e9ae51d43d6a96e303acdb0                         M   
1  58940d436a0d3d5da4e95a38                        XS   
2  5ea10e2c617580260d55310a                        XS   
3  5cedf17ad1d44c52c59e4aca                         M   
4  5ebf5819ebac372b070b0e70                         S   
5  5ec645bafd107a02279c8c54                         S   
6  5c63a337d5e2ce4bbb3152cf                        XS   
7  5e0ae5ebc2efb76ccf0a3391                         M   
8  5e6f1fb7fe4a5bb4b8bf36e5                         M   
9  5ccfaf238a8d535cec2dfb47                         S   

   product_variation_inventory  
0                           50  
1                           50  
2                            1  
3                           50  
4                            1  
5                            1  
6                           50  
7                           50  
8                           50  
9                           50

In [20]:
# load size_inventory into database 
size_inventory_df.to_sql(name='size_inventory', con=engine, if_exists='append', index=False)

8. Color Table    
The way to handle the color table is similar to the size_inventory table: we unify format, replace NULL values, drop duplicates, and rename columns. Since all color in the dataset are unique, we can create a new column with incrementing integer numbers for color_id. Next, we load the color table in database. 

In [21]:
# unify format 
colr = df['product_color'].value_counts()
colr[colr > 2]

df['product_color'] = df['product_color'].replace('Black', 'black')
df['product_color'] = df['product_color'].replace('coolblack', 'black')
df['product_color'] = df['product_color'].replace('offblack', 'black')

df['product_color'] = df['product_color'].replace('lightpurple', 'purple')

df['product_color'] = df['product_color'].replace('White', 'white')
df['product_color'] = df['product_color'].replace('whitefloral', 'white')
df['product_color'] = df['product_color'].replace('whitestripe', 'white')
df['product_color'] = df['product_color'].replace('ivory', 'white')


df['product_color'] = df['product_color'].replace('lightyellow', 'yellow')
df['product_color'] = df['product_color'].replace('beige', 'yellow')
df['product_color'] = df['product_color'].replace('apricot', 'yellow')
df['product_color'] = df['product_color'].replace('camel', 'yellow')

df['product_color'] = df['product_color'].replace('navyblue', 'blue')
df['product_color'] = df['product_color'].replace('lightblue', 'blue')
df['product_color'] = df['product_color'].replace('skyblue', 'blue')
df['product_color'] = df['product_color'].replace('darkblue', 'blue')
df['product_color'] = df['product_color'].replace('lakeblue', 'blue')
df['product_color'] = df['product_color'].replace('navy blue', 'blue')
df['product_color'] = df['product_color'].replace('navy', 'blue')
df['product_color'] = df['product_color'].replace('denimblue', 'blue')
df['product_color'] = df['product_color'].replace('prussianblue', 'blue')

df['product_color'] = df['product_color'].replace('winered', 'red')
df['product_color'] = df['product_color'].replace('rosered', 'red')
df['product_color'] = df['product_color'].replace('rose', 'red')
df['product_color'] = df['product_color'].replace('orange-red', 'red')
df['product_color'] = df['product_color'].replace('coralred', 'red')
df['product_color'] = df['product_color'].replace('Rose red', 'red')
df['product_color'] = df['product_color'].replace('burgundy', 'red')

df['product_color'] = df['product_color'].replace('lightpink', 'pink')
df['product_color'] = df['product_color'].replace('Pink', 'pink')

df['product_color'] = df['product_color'].replace('armygreen', 'green')
df['product_color'] = df['product_color'].replace('khaki', 'green')
df['product_color'] = df['product_color'].replace('lightgreen', 'green')
df['product_color'] = df['product_color'].replace('fluorescentgreen', 'green')
df['product_color'] = df['product_color'].replace('Army green', 'green')
df['product_color'] = df['product_color'].replace('mintgreen', 'green')
df['product_color'] = df['product_color'].replace('light green', 'green')
df['product_color'] = df['product_color'].replace('army green', 'green')
df['product_color'] = df['product_color'].replace('darkgreen', 'green')
df['product_color'] = df['product_color'].replace('army', 'green')
df['product_color'] = df['product_color'].replace('jasper', 'green')
df['product_color'] = df['product_color'].replace('lightkhaki', 'green')

df['product_color'] = df['product_color'].replace('gray', 'grey')
df['product_color'] = df['product_color'].replace('greysnakeskinprint', 'grey')
df['product_color'] = df['product_color'].replace('lightgray', 'grey')
df['product_color'] = df['product_color'].replace('lightgrey', 'grey')

df['product_color'] = df['product_color'].replace('coffee', 'brown')
df['product_color'] = df['product_color'].replace('tan', 'brown')

df['product_color'] = df['product_color'].replace('multicolor', 'other')
df['product_color'] = df['product_color'].replace('floral', 'other')
df['product_color'] = df['product_color'].replace('leopard', 'other')
df['product_color'] = df['product_color'].replace('camouflage', 'other')

df['product_color'] = df['product_color'].replace('white & green', 'dual')
df['product_color'] = df['product_color'].replace('black & green', 'dual')
df['product_color'] = df['product_color'].replace('black & white', 'dual')
df['product_color'] = df['product_color'].replace('red & blue', 'dual')
df['product_color'] = df['product_color'].replace('blue & pink', 'dual')
df['product_color'] = df['product_color'].replace('orange & camouflage', 'dual')
df['product_color'] = df['product_color'].replace('pink & black', 'dual')
df['product_color'] = df['product_color'].replace('pink & grey', 'dual')
df['product_color'] = df['product_color'].replace('white & black', 'dual')
df['product_color'] = df['product_color'].replace('pink & white', 'dual')
df['product_color'] = df['product_color'].replace('brown & yellow', 'dual')
df['product_color'] = df['product_color'].replace('navyblue & white', 'dual')
df['product_color'] = df['product_color'].replace('black & yellow', 'dual')
df['product_color'] = df['product_color'].replace('black & blue', 'dual')
df['product_color'] = df['product_color'].replace('pink & blue', 'dual')

df['product_color'] = df['product_color'].replace('camouflage', 'other')
df['product_color'] = df['product_color'].replace('RED', 'red')

df['product_color'] = df['product_color'].replace('rosegold', 'gold')

df['product_color'] = df['product_color'].replace(np.nan, 'other')
df['product_color'] = df['product_color'].replace('leopardprint', 'other')
df['product_color'] = df['product_color'].replace('star', 'other')
df['product_color'] = df['product_color'].replace('rainbow', 'other')
df['product_color'] = df['product_color'].replace('violet', 'other')

In [22]:
# create a dataframe 
color_df = pd.DataFrame(df['product_color'])
color_df = color_df.drop_duplicates(["product_color"])
color_df = color_df.rename(columns = {'product_color':'color_name'})

# assign unique id 
color_df.insert(0, 'color_id', range(1, 1 + len(color_df)))

In [23]:
# let's take a look 
color_df.head(100)

color_id color_name
0            1      white
1            2      green
2            3      other
3            4      black
4            5     yellow
5            6       blue
10           7       grey
12           8     orange
13           9        red
28          10       pink
47          11       dual
77          12     purple
139         13     silver
205         14      brown
1497        15       gold

In [24]:
# load color into database 
color_df.to_sql(name='color', con=engine, if_exists='append', index=False)

9. Product_currency   
We create a dataframe called "product_currency_df" by selecting "product_id","currency_buyer", and "currency_id"; rename "currency_buyer" with "currency_name". Then, we join "product_currency_df" and "currency_df" on "currency_name". Next, we modify column name and load the product_currency to database. 

In [25]:
# create dataframe 
product_currency_df = pd.DataFrame(df[['product_id','currency_buyer']])
product_currency_df = product_currency_df[['currency_buyer','product_id']]
product_currency_df = product_currency_df.rename(columns = {'currency_buyer':'currency_name'})
product_currency_df = pd.merge(product_currency_df,currency_df,on='currency_name')
product_currency_df = product_currency_df[['product_id', 'currency_id','currency_name']]

# let's take a look
product_currency_df.head()

product_id  currency_id currency_name
0  5e9ae51d43d6a96e303acdb0            1           EUR
1  58940d436a0d3d5da4e95a38            1           EUR
2  5ea10e2c617580260d55310a            1           EUR
3  5cedf17ad1d44c52c59e4aca            1           EUR
4  5ebf5819ebac372b070b0e70            1           EUR

In [26]:
# load product_currency into database 
product_currency_df = product_currency_df.rename(columns = {'currency_name':'currency_buyer'})
product_currency_df.to_sql(name='product_currency', con=engine, if_exists='append', index=False)

10. Product_shipping Table   
First, we create a dataframe called "product_shipping_df" by selecting "product_id", "shipping_option_name", "shipping_option_price", "shipping_is_express", and "countries_shipped_to". Then, we join "product_shipping_df" and "shipping_option_df" on "shipping_option_name". Next, we modify product_shipping_df and load product_shipping table to database. 

In [28]:
# create dataframe 
product_shipping_df = pd.DataFrame(df[['product_id','shipping_option_name','shipping_option_price', 
                                      'shipping_is_express','countries_shipped_to']])

# join table on 'shipping_option_name'
product_shipping_df = pd.merge(product_shipping_df,shipping_option_df, on='shipping_option_name')
product_shipping_df = product_shipping_df[['product_id','shipping_option_id','shipping_option_price', 
                                      'shipping_is_express','countries_shipped_to']]

In [29]:
# let's take a look 
product_shipping_df.head(2000)

product_id  shipping_option_id  shipping_option_price  \
0    5e9ae51d43d6a96e303acdb0                   1                      4   
1    58940d436a0d3d5da4e95a38                   1                      2   
2    5ea10e2c617580260d55310a                   1                      3   
3    5cedf17ad1d44c52c59e4aca                   1                      2   
4    5ebf5819ebac372b070b0e70                   1                      1   
..                        ...                 ...                    ...   
953  5c944cf5b82fb711ad500ec2                  12                      6   
954  5eb3937c9dd6ca23dcb12315                  14                      2   
955  5a1e957fce700d27f347c435                  14                      3   
956  588c54fb6661e84246a8aa71                  15                      7   
957  5e426615a4662e4a214c6371                  13                      3   

     shipping_is_express  countries_shipped_to  
0                      0                    34  
1                      0                    41  
2                      0                    36  
3                      0                    41  
4                      0                    35  
..                   ...                   ...  
953                    1                    43  
954                    0                    41  
955                    0                    39  
956                    1                    39  
957                    0                    46  

[958 rows x 5 columns]

In [31]:
# load product_shipping into database 
product_shipping_df.to_sql(name='product_shipping', con=engine, if_exists='append', index=False)

11. Product_color Table    
First, we rename the column "color_name" with "product_color". Then, we create a dataframe called "product_color_df" by selecting "product_id" and "product_color". Then we join "product_color_df" and "color_df" on "product_color". Last, we modify the dataframe and load product_shipping table in database. 

In [37]:
# rename column
color_df = color_df.rename(columns = {'color_name':'product_color'})

# create a dataframe 
product_color_df = pd.DataFrame(df[['product_id','product_color']])

# join table on 'product_color'
product_color_df = pd.merge(product_color_df,color_df, on='product_color')

# modify table 
product_color_df = product_color_df[['product_id','color_id','product_color']]

In [40]:
# let's take a look
product_color_df.head(1000)

product_id  color_id product_color
0    5e9ae51d43d6a96e303acdb0         1         white
1    5c63a337d5e2ce4bbb3152cf         1         white
2    5b07a7c6e997b33c6f9947da         1         white
3    5a4f2d8f5dcd0221008b2c36         1         white
4    5cc0049aca0c3b3d01a87964         1         white
..                        ...       ...           ...
953  5ea11b7fdb5ebf040c8e489d        14         brown
954  5e6f1aa7c182c5b14b486f9d        14         brown
955  5edefbd61eb25e17fedf9b8e        14         brown
956  5aa3592f6bcfa3795cba85b8        15          gold
957  5dbcfa073bab3307315950fc        15          gold

[958 rows x 3 columns]

In [41]:
# load product_color into database 
product_color_df.to_sql(name='product_color', con=engine, if_exists='append', index=False)

12. Tag Table    
First, we extract the column "tags" from df dataframe. Then we define a function to extract each single element from the lists of the tags column; apply the function and vertically concatenate each single tag element. Then, we drop the irrelevant column, rename the column to fit the designed table schema, and drop duplicated rows. Next, we nsert incrementing integers as tag_id into the first column and load tag table in database. 


In [86]:
# create dataframe
tag_df = pd.DataFrame(df['tags'])

# define a function
def fun(row):
    elements = row['tags']
    elements = elements.split(',')
    out = np.array([row] * len(elements))
    out['tags'] = elements
    
# apply the function and vertically concatenate each single tag element    
tag_df = pd.concat([pd.Series(row['tags'].split(','))              
                    for _, row in tag_df.iterrows()]).reset_index()

# drop the  irrelevant column.
tag_df = tag_df.drop('index', 1)

# rename the column to fit the designed table schema.
tag_df = tag_df.rename(columns = {0:'tag_name'})

tag_df.head(100)

tag_name
0               Summer
1              Fashion
2   womenunderwearsuit
3    printedpajamasset
4     womencasualshort
..                 ...
95          sleeveless
96               Women
97              Casual
98            jumpsuit
99              blouse

[100 rows x 1 columns]

In [87]:
# insert incrementing integers as tag_id into the first column

tag_df = tag_df.drop_duplicates()
tag_df.insert(0, 'tag_id', range(1, 1 + len(tag_df)))

# let's take a look
tag_df.head()

tag_id            tag_name
0       1              Summer
1       2             Fashion
2       3  womenunderwearsuit
3       4   printedpajamasset
4       5    womencasualshort

In [88]:
# load tag into database 
tag_df.to_sql(name='tag', con=engine, if_exists='append', index=False)

13. Product_tag Table    
First, we extract "product_id" and "tags" columns from df dataframe and apply the function that created previously to extract each single element from the lists of the tags column and then vertically concatenate each single tag element. Then we change the order of the column, rename the column, and drop duplicated rows. Next, we join "product_tag_df" and "tag_df" on "tags". Last, we change the order of the columns to fit the order of the designed table schema and load to database. 

In [106]:
# extract the columns above  from df dataframe
product_tag_df = pd.DataFrame(df[['product_id','tags']]) 

# apply the function that is created on 12 to extract each single element from the lists of the tags column and then vertically concatenate each single tag element.
product_tag_df = pd.concat([pd.Series(row['product_id'], row['tags'].split(','))              
                    for _, row in product_tag_df.iterrows()]).reset_index()
# change the order of the columns
product_tag_df = product_tag_df[[0, 'index']]

# rename the column to fit the designed table schema
product_tag_df = product_tag_df.rename(columns = {0:'product_id'})

# rename the column to fit the designed table schema
product_tag_df = product_tag_df.rename(columns = {'index':'tags'})

# drop duplicated rows
product_tag_df = product_tag_df.drop_duplicates()

In [111]:
# rename the column of tag_df for better performing inner join later
tag_df = tag_df.rename(columns = {'tag_name':'tags'})

# inner join table product_tag_df and table tag_df 
product_tag_df = pd.merge(product_tag_df,tag_df, on='tags')

# change the order of the columns to fit the order of the designed table schema
product_tag_df = product_tag_df[['product_id','tag_id','tags']]

In [117]:
# let's take a look 
product_tag_df.head(10000)

product_id  tag_id           tags
0     5e9ae51d43d6a96e303acdb0       1         Summer
1     58940d436a0d3d5da4e95a38       1         Summer
2     5ea10e2c617580260d55310a       1         Summer
3     5cedf17ad1d44c52c59e4aca       1         Summer
4     5ebf5819ebac372b070b0e70       1         Summer
...                        ...     ...            ...
9995  5e3a58fbab71890480516a09     133  topsamptshirt
9996  5c837b60bec7771428a98865     133  topsamptshirt
9997  5d63a7dfa9c84677ea42887f     133  topsamptshirt
9998  5d2fe4d55cdae96941f5a9bd     133  topsamptshirt
9999  5a618aff64621a6a2122adb6     133  topsamptshirt

[10000 rows x 3 columns]

In [120]:
# load product_tag into database 
product_tag_df.to_sql(name='product_tag', con=engine, if_exists='append', index=False)

14. Badge Table   
First, we manually create a dataframe called "badge_df". Then we rename the columns to fit the designed table schema. Last, we load the badge table to database.  

In [122]:
# manually create the badge_df
badge_df = [[1, 'badge_local_product'], [2, 'badge_product_quality'], [3, 'badge_fast_shipping']]

# rename the columns to fit the designed table schema
badge_df = pd.DataFrame(badge_df, columns = ['badge_id', 'badge_name']) 

# let's take a look
badge_df.head()

badge_id             badge_name
0         1    badge_local_product
1         2  badge_product_quality
2         3    badge_fast_shipping

In [123]:
# load badge into database 
badge_df.to_sql(name='badge', con=engine, if_exists='append', index=False)

15. Product_badge Table   
First, we extract columns "product_id", "badge_local_product", "badge_product_quality", and "badge_fast_shipping" from df dataframe. Then we create three new columns "badge_name_1", "badge_name_2", and "badge_name_3"; drop unnecessary columns. Then, we create a new column ‘badge_name’ to merge the information from column ‘badge_name_1’, ‘badge_name_2’, and ‘badge_name_3’. Next, we apply the function that created in table 12 to extract each single element from the lists of the‘badge_name’ column and then vertically concatenate each single element; change column name;inner join table product_bage_df and table badge_df. Last, we load product_badge table to database. 


In [367]:
# extract the columns above from df dataframe
product_badge_df = pd.DataFrame(df[['product_id','badge_local_product','badge_product_quality','badge_fast_shipping']]) 

In [368]:
# create a new column ‘badge_name_1’; if ‘badge_local_product’ == 0, assign None to column ‘badge_name_1’; if if ‘badge_local_product’ == 1, assign ‘badge_local_product’ to column ‘badge_name_1’
product_badge_df.loc[product_badge_df['badge_local_product'] == 0, 'badge_name_1'] = None
product_badge_df.loc[product_badge_df['badge_local_product'] == 1, 'badge_name_1'] = 'badge_local_product'

# create a new column ‘badge_name_2’; if ‘badge_product_quality’ == 0, assign None to column ‘badge_name_2’; if if ‘badge_product_quality’ == 1, assign ‘badge_product_quality’ to column ‘badge_name_2’
product_badge_df.loc[product_badge_df['badge_product_quality'] == 0, 'badge_name_2'] = None
product_badge_df.loc[product_badge_df['badge_product_quality'] == 1, 'badge_name_2'] = 'badge_product_quality' 

# create a new column ‘badge_name_3’; if ‘badge_fast_shipping’ == 0, assign None to column ‘badge_name_3’; if if ‘badge_fast_shipping’ == 1, assign ‘badge_fast_shipping’ to column ‘badge_name_3’
product_badge_df.loc[product_badge_df['badge_fast_shipping'] == 0, 'badge_name_3'] = None
product_badge_df.loc[product_badge_df['badge_fast_shipping'] == 1, 'badge_name_3'] = 'badge_fast_shipping'

In [369]:
# drop the following columns
product_badge_df = product_badge_df.drop('badge_local_product', 1)
product_badge_df = product_badge_df.drop('badge_product_quality', 1)
product_badge_df = product_badge_df.drop('badge_fast_shipping', 1)

In [370]:
# create a new column ‘badge_name’ to merge the information from column ‘badge_name_1’, ‘badge_name_2’, and ‘badge_name_3’
product_badge_df['badge_name'] = product_badge_df[product_badge_df.columns[1:]].apply(
    lambda x: ','.join(x.dropna().astype(str)),
    axis=1
)

# drop column ‘badge_name_1’, ‘badge_name_2’, and ‘badge_name_3’
product_badge_df = product_badge_df.drop('badge_name_1', 1)
product_badge_df = product_badge_df.drop('badge_name_2', 1)
product_badge_df = product_badge_df.drop('badge_name_3', 1)

In [371]:
# extract the rows that don’t have None value
product_badge_df = product_badge_df[product_badge_df.badge_name != '']

In [372]:
# apply the function that is created on 12 to extract each single element from the lists of the‘badge_name’ column and then vertically concatenate each single element
product_badge_df = pd.concat([pd.Series(row['product_id'], row['badge_name'].split(','))              
                    for _, row in product_badge_df.iterrows()]).reset_index() 

# change the order of the columns to fit the order of the designed table schema
product_badge_df = product_badge_df[[0, 'index']]

# rename the columns to fit the designed table schema
product_badge_df = product_badge_df.rename(columns = {0:'product_id'})

# rename the column of for better performing inner join later
product_badge_df = product_badge_df.rename(columns = {'index':'badge_name'}) 

product_badge_df.head()

product_id             badge_name
0  5a4f2d8f5dcd0221008b2c36  badge_product_quality
1  5c9ee457c233fd74ef4c8a84  badge_product_quality
2  5b10cf0b7cc4492c87411045  badge_product_quality
3  556fbe8595528719e004fa27  badge_product_quality
4  57d3bc906cfeb9135f44ded5    badge_local_product

In [373]:
# inner join table product_bage_df and table badge_df 
product_badge_df = pd.merge(product_badge_df,badge_df, on='badge_name')

# drop the irrelevant column
product_badge_df = product_badge_df.drop('badge_name', 1)

In [374]:
# let's take a look
product_badge_df.head()

product_id  badge_id
0  5a4f2d8f5dcd0221008b2c36         2
1  5c9ee457c233fd74ef4c8a84         2
2  5b10cf0b7cc4492c87411045         2
3  556fbe8595528719e004fa27         2
4  5d78eb19cc404c03393c83b5         2

In [376]:
# load product_badge into database 
product_badge_df.to_sql(name='product_badge', con=engine, if_exists='append', index=False)